### PR TITLE
refactor(message-sending): Introduce an abstraction for making requests WPB-5150 WPB-5153 #2

### PR DESCRIPTION
**Pull request by typfel**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5150" title="WPB-5150" target="_blank">Sub-taskWPB-5150</a>  [iOS] Extract session handling into separate component (SessionEstablisher)
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- Introduce an abstraction for making requests (fetching prekeys, sending proteus messages), which hides the API differences between the different api versions.
- Don't rely on `missedClients` relationship on the self client for establishing missing sessions, instead we establish sessions only with clients which reported missing by the backend.

### Testing

I'll add tests in a another PR when the class boundaries are stable

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift [hunk 1]
```diff
@@ -117,124 +111,120 @@ public class MessageSender {
...
             }
         }
     }
 
-    private func missingClients() -> Set<UserClient> {
+    private func missingClients() -> Set<QualifiedClientID> {
         return managedObjectContext.performAndWait {
-            ZMUser.selfUser(in: managedObjectContext).selfClient()?.missingClients ?? Set()
+            let clients = ZMUser.selfUser(in: managedObjectContext).selfClient()?.missingClients ?? Set()
+            return Set(clients.compactMap({ $0.qualifiedClientID })) // FIXME: [jacob] we can't do compact map
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift:154]: is this still todo?

File: wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift [hunk 2]
```diff
@@ -117,124 +111,120 @@ public class MessageSender {
...
+            default:
+                return .failure(MessageSendError.networkError(failure))
             }
-
-            if case .permanentError = response.result {
-                WireLogger.messaging.warn("got \(response.httpStatus), not retrying")
-                return .success(false)
+        default:
+            if case .permanentError = failure.response?.result {
+                return .success(Set()) // FIXME: [jacob] it's dangerous to retry indefinitely like this
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift]: done in another PR?

File: wire-ios-request-strategy/Sources/Message Sending/SessionEstablisher.swift [hunk 3]
```diff
@@ -19,29 +19,30 @@
...
+        self.apiProvider = apiProvider
         self.managedObjectContext = context
     }
 
-    private let httpClient: HttpClient
+    private let apiProvider: APIProvider
     private let managedObjectContext: NSManagedObjectContext
     private let requestFactory = MissingClientsRequestFactory()
     private let processor = PrekeyPayloadProcessor()
+    private let BATCH_SIZE = 28
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Message Sending/SessionEstablisher.swift]:
```suggestion
    private let batchSize = 28
```

File: wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift (hunk 2 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift]: No, I've kept this I'm keeping the existing behaviour but I think we should re-evaluate how we want to approach this.

File: wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift (hunk 1 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift:154]: No, it got solved in https://github.com/wireapp/wire-ios/pull/622

File: wire-ios-request-strategy/Sources/Message Sending/SessionEstablisher.swift (hunk 3 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Message Sending/SessionEstablisher.swift]: Done

File: wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift (hunk 2 above)
Comment (netbe) [reply wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift]: ok, let's have a ticket for it then if we don't solve it now in the parent ticket

Comment (typfel) [reply wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift]: https://wearezeta.atlassian.net/browse/WPB-5454

**Review by netbe — APPROVED** [review]

**Review by johnxnguyen — APPROVED** [review]
